src/DiagnosticsManager.ts

```diff
@@ -0,0 +1,281 @@
+     */
+    handleDiagnostics(uri: vscode.Uri, sources: string[], diagnostics: vscode.Diagnostic[]): void {
+        // Remove the old set of benefits from this source
+        const currentDiagnostics = this.removeDiagnostics(
+            this.diagnosticCollection.get(uri)?.slice() || [],
+            sources
+        );
+        let newDiagnostics: vscode.Diagnostic[];
+        switch (configuration.diagnosticsCollection) {
+            case "keepSourceKit":
+                newDiagnostics = this.mergeDiagnostics(
+                    currentDiagnostics,
+                    diagnostics,
+                    DiagnosticsManager.sourcekit
+                );
+                break;
+            case "keepSwiftc":
+                newDiagnostics = this.mergeDiagnostics(
+                    currentDiagnostics,
+                    diagnostics,
+                    DiagnosticsManager.swiftc
+                );
+                break;
+            case "onlySourceKit":
+                newDiagnostics = this.removeDiagnostics(
+                    currentDiagnostics,
+                    DiagnosticsManager.swiftc
+                ); // Just in case
+                if (DiagnosticsManager.swiftc.find(s => sources.includes(s))) {
+                    break;
+                }
+                newDiagnostics.push(...diagnostics);
+                break;
+            case "onlySwiftc":
+                newDiagnostics = this.removeDiagnostics(
+                    currentDiagnostics,
+                    DiagnosticsManager.sourcekit
+                ); // Just in case
+                if (DiagnosticsManager.sourcekit.find(s => sources.includes(s))) {
+                    break;
+                }
+                newDiagnostics.push(...diagnostics);
+                break;
+            case "keepAll":
+                newDiagnostics = currentDiagnostics.concat(diagnostics);
+                break;
+        }
+        this.diagnosticCollection.set(uri, newDiagnostics);
+    }
+
+    private mergeDiagnostics(
+        currentDiagnostics: vscode.Diagnostic[],
+        newDiagnostics: vscode.Diagnostic[],
+        precedence: string[]
+    ): vscode.Diagnostic[] {
+        const merged: vscode.Diagnostic[] = [];
+        const messages: string[] = [];
+        for (const diagnostic of newDiagnostics) {
+            // De-duplicate duplicate diagnostics from Swift
+            // TODO remove when https://github.com/apple/swift/issues/73973 is fixed
+            if (messages.includes(diagnostic.message)) {
+                continue;
+            }
+            messages.push(diagnostic.message);
+            // See if a duplicate diagnostic exists
+            const currentDiagnostic = currentDiagnostics.find(
+                d =>
+                    d.range.start.isEqual(diagnostic.range.start) &&
+                    d.message === diagnostic.message
+            );
```

Comment:
As I mentioned in Slack. Swift 5.10 compiler produces diagnostics with the first character capitalised and SourceKit-LSP produces diagnostics with everything lowercase. Thus nothing is merged